src/compose/app.ts

```diff
@@ -698,6 +679,27 @@ export class App {
 		}
 	}
 
+	private dependenciesMetForServiceFetch(
+		servicePairs: Array<ChangingPair<Service>>,
+	) {
+		// Current services should be killed before target
+		// service fetch depending on update strategy
+		const servicePairsWithCurrent = servicePairs.filter(
+			(pair) => pair.current != null,
+		);
+		for (const { current, target } of servicePairsWithCurrent) {
+			// Prefer target's update strategy if target service exists
+			const strategy = getStrategyFromService(target ?? current!);
+			if (
```

Comment:
I understood what you meant by this code when reading the tests, I think this is wrong as this means we cannot fetch any service image from the new until all services with (kill|delete)-then-download from the previous release have been stopped|removed. Release strategies are still per-service with this change, but we want to chose when to apply changes when it's safest for the app (meaning less downtime). With this particular code, it could be that the new images take a long time to download, meaning the app is unusable during that whole time.

Reply:
By this, do you mean that we should not adhere strictly to the concept of app-wide strategies, but rather make allowances that benefit performance, such as starting a fetch for a target service even when current services haven't yet been killed while using `kill|delete-then-download`?

Reply:
Hmm, the term ''app wide strategies" is a bit ambiguous, we might need to look for a new term. The strategies are still per service, as they define the order of operations for updating a single service, but when updating the app, we need to combine those service operations in a way that minimizes downtime to the app, and allows us to rollback quickly if something goes wrong with the update. 

This means, suppose we have a combination of 'download-then-kill', 'kill-then-download' and 'delete-then-download' services,

1. Pull all new `download-then-kill` target images
2. Kill all `kill-then-download` and `delete-then-download` services the are being updated and remove images for the latter (in a future improvement we'll also need to kill services that depend on these)
3. Download the new images for the `kill-then-download` and `delete-then-download` services.
4. Kill remaining services to be updated and start all new services
5. Cleanup unused images

This also means, that if we need to roll back before step 5, we can do it pretty quickly (except for re-downloading the `delete-then-download`)

Let me know if this makes sense

Reply:
Actually, on second thought, your PR swaps steps 1 and 2, which is OK for now. We will want to move to the above sequence when we use `depends-on` for kill, so we kill services as late as possible

Reply:
Would step 1 also apply for new target images with `kill|delete-then-download` ?